File: extended/logistic-regression/gradient-imperative/Gradient.scala

import java.util.Random

object Gradient {

  case class Sample(vec: Array[Double], label: Double)

  // A psuedo random function.
  def prf(input: Long) : Double = {
    var x = input + 4698;
    x = ((x >> 16) ^ x) * 0x45d9f3b;
    x = ((x >> 16) ^ x) * 0x45d9f3b;
    x = ((x >> 16) ^ x);
    return x;
  }

  def time[R](block: => R): R = {
    val t0 = System.nanoTime()
    val result = block    // call-by-name
    val t1 = System.nanoTime()
    println("Elapsed time: " + (t1 - t0)/1e6 + "ms")
    result
  }

  def vector_add(x: Array[Double], y: Array[Double]) : Array[Double] = {
    return Array.tabulate(x.length){i => x(i) + y(i)};
  }

  def vector_sub(x: Array[Double], y: Array[Double]) : Array[Double] = {
    return Array.tabulate(x.length){i => x(i) - y(i)};
  }

  def vector_dot(x: Array[Double], y: Array[Double]) : Double = {
    var result: Double = 0
      for (i <- 0 until x.length)
        result += x(i) * y(i)
    return result
  }

  def vector_mul(x: Double, y:Array[Double]) : Array[Double] = {
    return Array.tabulate(y.length){i => y(i) * x}
  }

  def print_help() {
    println("Usage: pass 3 arguments")
    println("   <Int dimension>")
    println("   <Int iteration_num>")
    println("   <Float sample_num in million>")
  }

  def main(args: Array[String]) {
    var (kDimension          : Int,
         kIterationNum       : Int,
         kSampleNumInMillion : Float) =
    args.length match {
      case 3 => (args(0).toInt, args(1).toInt, args(2).toFloat)
      case default => print_help();
    }

    var kSampleNum          : Int = (kSampleNumInMillion.toFloat * 1000000).toInt

    println("** dimension    : " + kDimension)
    println("** iteration num: " + kIterationNum)
    println("** sample num   : " + kSampleNum)

    var samples = Array.tabulate(0)(x => Sample(Array.tabulate(0)(x => 0), 0))
    println("\nSamples generation: ")
    time{
      samples = Array.tabulate(kSampleNum)(x => x)
        .map(x => {
                    val label = if (x % 2 == 0) -1 else 1
                    val vec = Array.tabulate(kDimension)(y => prf(x*kDimension+y))
                    Sample(vec, label)
                  })
    }

    // initialize w to a random vector
    var rand = new Random(42)
    var w: Array[Double] = Array.tabulate(kDimension)(i => rand.nextDouble)

    var elapsed_list : Array[Double] = Array.tabulate(0)(x => 0)
    var truncate = 10;

    for (i <- 1 to kIterationNum) {
      val t0 = System.nanoTime()
      val gradient = samples
        .map(p => {
                    val factor =  (1 / (1 + math.exp(p.label * (vector_dot(w,p.vec)))) - 1) * p.label;
                    vector_mul(factor, p.vec)
                  })
        .reduce(vector_add)
        w = vector_add(w, gradient)

      val t1 = System.nanoTime()
      val elapsed = (t1 - t0)/1e6
      println("Iteration " + i + " elapsed time: " + elapsed + "(ms)")
      if (i > truncate) {
        elapsed_list = elapsed_list :+ elapsed
      }
    }

    val average = elapsed_list.reduce((x, y) => x + y) / elapsed_list.length
    println("Average for " + elapsed_list.length + " iterations (excluding the first " + truncate +  "): " + average + "(ms)")
  }
}

File: extended/logistic-regression/gradient-java/Gradient.java

import java.util.List;
import java.util.Iterator;
import java.util.ArrayList;
import java.util.Arrays;
import java.lang.Math;

class Sample {
  double label_;
  double[] vector_;
}

class Gradient {

  public static void VectorAddWithScale(double[] acc, double[] add, double scale) {
    assert acc.length == add.length;
    int length = acc.length;

    for (int i = 0; i < length; ++i) {
      acc[i] = acc[i] + scale * add[i];
    }
  }

  public static double VectorDot(double[] v1, double[] v2) {
    assert v1.length == v2.length;
    int length = v1.length;

    double result = 0;
    for (int i = 0; i < length; ++i) {
      result += v1[i] * v2[i];
    }
    return result;
  }


  public static void PrintHelp() {
    System.out.println("Usage: pass 3 arguments");
    System.out.println("   <Int dimension>");
    System.out.println("   <Int iteration_num>");
    System.out.println("   <Float sample_num in million>");
  }

  public static void main(String[] args) {

    Integer dimension = 10;
    Integer iteration_num = 20;
    Double sample_num_m = 1.0;
    if (args.length != 3) {
      System.out.println("Error!");
      PrintHelp();
      return;
    } else {
      dimension = Integer.parseInt(args[0]);
      iteration_num = Integer.parseInt(args[1]);
      sample_num_m = Double.parseDouble(args[2]);
    }

    Integer sample_num = (Integer)(int)(sample_num_m * 1e6);

    System.out.println("dimension     : " + String.valueOf(dimension));
    System.out.println("iteration num : " + String.valueOf(iteration_num));
    System.out.println("sample num    : " + String.valueOf(sample_num));


    Sample[] samples = new Sample[sample_num];
    {
      long start_time = System.nanoTime();
      for (int i = 0; i < sample_num; ++i) {
        samples[i] = new Sample();
        samples[i].label_ = (double)(i % 2) * 2 - 1;
        samples[i].vector_ = new double[dimension];
        for (int j = 0; j < dimension; ++j) {
          samples[i].vector_[j] = 13.0;
        }
      }
      long end_time = System.nanoTime();
      System.out.println("Sample generation elapsed time : " + String.valueOf((end_time - start_time)/1e6) + " ms");
    }

    double[] weight = new double[dimension];
    for (int j = 0; j < dimension; ++j) {
      weight[j] = 1;
    }

    List<Double> elapsed_list = new ArrayList<Double>();
    int truncate = 10;

    for (int iter = 0; iter < iteration_num; ++iter) {
      long start_time = System.nanoTime();
      double[] gradient = new double[dimension];
      for (int j = 0; j < dimension; ++j) {
        gradient[j] = 0;
      }

      for (int i = 0; i < sample_num; ++i) {
        Sample s = samples[i];
        double scale =  (1.0 / (1.0 + Math.exp(s.label_ * (VectorDot(weight, s.vector_)))) - 1) * s.label_;
        VectorAddWithScale(gradient, s.vector_, scale);
      }
      VectorAddWithScale(weight, gradient, 1.0);

      long end_time = System.nanoTime();
      double elapsed = (end_time - start_time)/1e6;
      System.out.println("Iteration " + iter + " elapsed time : " + String.valueOf(elapsed) + "(ms)");
      if (iter >= truncate) {
        elapsed_list.add(elapsed);
      }
    }

    Double average = 0.0;
    {
      Iterator<Double> iter = elapsed_list.iterator();
      while(iter.hasNext()) {
        average += iter.next();
      }
    }
    average = average / elapsed_list.size();
    System.out.println("Average gradient for " + iteration_num + " iterations (excluding the first " + String.valueOf(truncate) + "): " + String.valueOf(average) + "(ms)");
  }
}


